src/resdk/resources/metadata.py

```diff
@@ -0,0 +1,232 @@
+        if not intersection:
+            warnings.warn(
+                "No intersection between samples in df and samples in collection."
+            )
+
+        not_in_col = df_samples - col_samples
+        if not_in_col:
+            not_in_col = ", ".join(list(not_in_col))
+            warnings.warn(
+                f"There are samples in df that are not in collection: {not_in_col}"
+            )
+
+        not_in_df = col_samples - df_samples
+        if not_in_df:
+            not_in_df = ", ".join(list(not_in_df))
+            warnings.warn(
+                f"There are samples in collection that are not in df: {not_in_df}"
+            )
+
+    def get_df(self, parser=None, **kwargs):
+        """Get table as pd.DataFrame."""
+        # Do not use cached value if parser is specified
+        if self._df is None or parser is not None:
+
+            # Enable parsing the byte stream with arbitrary parser, not just pandas
+            # Otherwise try to guess the parser based on file extension
+            basename = self.output["table"]["file"]
+            if parser is not None:
+                df = parser(self.df_bytes, **kwargs)
+            elif basename.endswith("xls"):
+                df = pd.read_excel(self.df_bytes, engine="xlrd")
+            elif basename.endswith("xlsx"):
+                df = pd.read_excel(self.df_bytes, engine="openpyxl")
+            elif any(basename.endswith(ext) for ext in ["tab", "tsv"]):
+                df = pd.read_csv(self.df_bytes, sep="\t")
+            else:
+                df = pd.read_csv(self.df_bytes)
+
+            self.validate_df(df)
+            self._df = df
+
+        return self._df
+
+    def set_df(self, value):
+        """Set df."""
+        if self.id:
+            raise ValueError(
+                "Setting df attribute on already uploaded Metadata is not allowed."
+            )
+        if not self.collection:
+            # Validation is not possible without collection
+            raise ValueError(
+                "Setting df attribute before setting collection is not allowed."
+            )
+
+        self.validate_df(value)
+        self._df = value
+
+    df = property(get_df, set_df)
+
+    def save(self):
+        """Save Metadata to the server.
+
+        If Metadata is already uploaded: update. Otherwise, create new one.
+        """
+        if self.id:
+            super().save()
+        else:
+            if self.df is None or self.df.empty:
+                raise ValueError("Attribute df must be set before saving.")
+            if not self.collection:
+                raise ValueError("Collection must be set before saving.")
+
+            # All resdk machinery for uploading files works with real
+            # files on the system. Ideally we would support "file upload"
+            # from a stream, but for now, let's use a tempfile solution.
+            tmp_file = tempfile.NamedTemporaryFile(suffix=".csv")
+            self.df.to_csv(tmp_file.name)

```

Comment:
Would it make sense to skip writing the index here, since DataFrame index is not being set to meaningful values?